lib/services/security/permissions.js

```diff
@@ -0,0 +1,282 @@
+const CONSTANTS = require('../..').constants;
+var ALLOWED_PERMISSIONS = [
+  'set',
+  'get',
+  'remove',
+  'on',
+  '*',
+  'delete',
+  'put',
+  'post',
+  'head',
+  'options'
+];
+
+module.exports = class Permissions {
+  constructor(config, type, happn, securityService) {
+    this.__config = this.defaults(config);
+    this.securityService = securityService;
+    this.cacheService = happn.services.cache;
+    this.dataService = happn.services.data;
+    this.type = type;
+    this.cache = this.cacheService.new(`cache_security_${this.type}_permissions`, {
+      type: 'LRU',
+      cache: this.__config.__cache_permissions
+    });
+    this.__userPrefix = this.__config.__userPermissionsPrefix;
+  }
+
+  static create(config, type, happn, securityService) {
+    return new Permissions(config, type, happn, securityService);
+  }
+
+  defaults(config) {
+    let defaultConfig = !config ? {} : { ...config };
+    if (!defaultConfig.__cache_permissions)
+      defaultConfig.__cache_permissions = {
+        max: 10000,
+        maxAge: 0
+      };
+    defaultConfig.__userPermissionsPrefix = defaultConfig.__userPermissionsPrefix || '_USER/';
+    return defaultConfig;
+  }
+
+  attachPermissions(entity) {
+    return new Promise((resolve, reject) => {
+      entity.permissions = {};
+
+      this.listPermissions(entity.name || entity.username)
+        .then(permissions => {
+          permissions.forEach(permission => {
+            if (permission.authorized) {
+              if (!entity.permissions[permission.path])
+                entity.permissions[permission.path] = { actions: [] };
+              entity.permissions[permission.path].actions.push(permission.action);
+            }
+          });
+          resolve(entity);
+        })
+        .catch(e => {
+          reject(e);
+        });
+    });
```

Comment:
Should return from then() instead of resolving a new promise from an existing promise